Retroupper can't be undone with spaces after

## Summary

While working on my emoji dictionary, I noticed some weird behavior when trying to undo retroactions with spaces after. I found the same behavior with retroupper:

## Reproducing

```diff
diff --git a/test/test_blackbox.py b/test/test_blackbox.py
index 731c5d9..d32f245 100644
--- a/test/test_blackbox.py
+++ b/test/test_blackbox.py
@@ -912,6 +912,27 @@ class TestsBlackbox(BlackboxTester):
         TPAO/AES/KPATS " Foo's"
         '''

+    def test_retro_upper19(self): # WORKING
+        r'''
+        "TPAO": "foo",
+        "KPATS": "{*<}",
+
+        TPAO " foo"
+        KPATS " FOO"
+        * " foo"
+        '''
+
+    def test_retro_upper20(self): # FAILS
+        r'''
+        "TPAO": "foo",
+        "KPATS": "{*<}",
+
+        :spaces_after
+        TPAO "foo "
+        KPATS "FOO "
+        * "foo "
+        '''
+
     def test_upper1(self):
         r'''
         'TP*U': '{<}',
```

Result:

```
================================================ FAILURES =================================================
____________________________________ TestsBlackbox.test_retro_upper20 _____________________________________

bb = <test.test_blackbox.TestsBlackbox object at 0x1233c4da0>

>   new_f = lambda bb: (f(bb), replay(bb, name, test))
E   AssertionError: test_retro_upper20
E
E     "TPAO": "foo",
E     "KPATS": "{*<}",
E
E     :spaces_after
E     TPAO "foo "
E     KPATS "FOO "
E   > * "foo "
E
E      ' '
E   != 'foo '

plover_build_utils/testing.py:118: AssertionError
================================== 1 failed, 489 passed in 4.59 seconds ===================================
```

## Plover Version

314a61f

--- a/plover/formatting.py
+++ b/plover/formatting.py
@@ -453,8 +453,14 @@ class OutputHelper:
         # 2
         # >>> len(unicodedata.normalize('NFC', u"C\u0327"))
         # 1
-        after = self.after.appended_text
-        before = self.after.replaced_text + self.before.appended_text
+        if len(self.before.replaced_text) > len(self.after.replaced_text):
+            assert self.before.replaced_text.endswith(self.after.replaced_text)
+            replaced_text = self.before.replaced_text
+        else:
+            assert self.after.replaced_text.endswith(self.before.replaced_text)
+            replaced_text = self.after.replaced_text
+        before = replaced_text[:len(replaced_text)-len(self.before.replaced_text)] + self.before.appended_text
+        after = replaced_text[:len(replaced_text)-len(self.after.replaced_text)] + self.after.appended_text
         common_length = len(commonprefix([before, after]))
         erased = len(before) - common_length
         if erased:
@@ -470,8 +476,6 @@ class OutputHelper:
         for action in self.before.render(undo, last_action):
             pass
         # Render new actions.
-        if self.before.replaced_text:
-            self.after.appended_text = self.before.replaced_text
         for action in self.after.render(do, last_action):
             self.flush()
             if action.combo:

--- a/test/test_blackbox.py
+++ b/test/test_blackbox.py
@@ -912,6 +912,27 @@ class TestsBlackbox(BlackboxTester):
         TPAO/AES/KPATS " Foo's"
         '''
 
+    def test_retro_upper19(self):
+        r'''
+        "TPAO": "foo",
+        "KPATS": "{*<}",
+
+        TPAO " foo"
+        KPATS " FOO"
+        * " foo"
+        '''
+
+    def test_retro_upper20(self):
+        r'''
+        "TPAO": "foo",
+        "KPATS": "{*<}",
+
+        :spaces_after
+        TPAO "foo "
+        KPATS "FOO "
+        * "foo "
+        '''
+
     def test_upper1(self):
         r'''
         'TP*U': '{<}',
@@ -1348,4 +1369,14 @@ class TestsBlackbox(BlackboxTester):
 
         TEFT/W-G  ' TESTING\nwith'
         W-G       ' TESTING\nwithing\nwith'
+        '''
+
+    def test_bug961(self):
+        r'''
+        "PHAED": "made",
+        "KWREUS": "{^ies}",
+
+        :spaces_after
+        PHAED/KWREUS  'madies '
+        *             'made '
         '''

